File: Java/MinimumFallingPathSumII.java
import java.util.*;

class Solution {
    public int solve(int row, int col, int[][] matrix, int[][] dp) {

        if(row == matrix.length - 1) 
            return matrix[row][col];

        if(dp[row][col] != -1) return dp[row][col];

        int min = Integer.MAX_VALUE;
        for(int c = 0; c < matrix[0].length; c++) {
            if(c == col) continue;
            min = Math.min(min, matrix[row][col] + solve(row + 1, c, matrix, dp));
        }

        dp[row][col] = min;
        return dp[row][col];

    }
    public int minFallingPathSum(int[][] matrix) {
        
        int n = matrix.length;
        int minimumSum = Integer.MAX_VALUE;

        int[][] dp = new int[n][n];
        for(int d[] : dp) {
            Arrays.fill(d, -1);
        }

        for(int col = 0; col < n; col++) {
            int currentSum = solve(0, col, matrix, dp);
            minimumSum = Math.min(minimumSum, currentSum);
        }

        return minimumSum;
    }
}

File: CPP/MinimumFallingPathSumII.cpp
class Solution {
public:

    int solve(int row, int col, vector<vector<int>>& grid, vector<vector<int>>& dp) {
        if(row == grid.size() - 1) return grid[row][col];

        if(dp[row][col] != -1) return dp[row][col];

        int minimumSum = INT_MAX;
        for(int c = 0; c < grid[0].size(); c++) {
            if(c != col) {
                minimumSum = min(minimumSum, grid[row][col] + solve(row + 1, c, grid, dp));
            }
        }
        return dp[row][col] = minimumSum;
    }

    int minFallingPathSum(vector<vector<int>>& grid) {
        int n = grid.size();
        int minimumSum = INT_MAX;
        vector<vector<int>> dp(n, vector<int>(n, -1));

        for(int i = 0; i < n; i++) {
            minimumSum = min(minimumSum, solve(0, i, grid, dp));
        }
        return minimumSum;
        
    }
};